[integration] Replace gLogger by the standard library: logging

aldbr opened the pull request:
In this branch, there are: 
- changes, mainly in _FrameworkSystem/private/standardLogging_ : the new solution based on logging
- tests in _FrameworkSystem/tests,_ and there are also some "tests" in a temporary system named : _TestLoggerSystem_

aldbr committed:
starting tests on gLogger and Logging

aldbr committed:
starting tests on gLogger and Logging

aldbr committed:
add new system tests for gLogger

aldbr committed:
add new system tests for gLogger

aldbr committed:
improve system tests for gLogger

aldbr committed:
improve system tests for gLogger

aldbr committed:
delete some useless files

aldbr committed:
delete some useless files

aldbr committed:
add wrapper logging class and configuration logging class

add custom formatter with system/component name

aldbr committed:
add wrapper logging class and configuration logging class

add custom formatter with system/component name

aldbr committed:
add some lines to operate logging system for clients and agents

aldbr committed:
add some lines to operate logging system for clients and agents

aldbr committed:
add some lines to operate logging system for scripts

first test on color logs

aldbr committed:
add some lines to operate logging system for scripts

first test on color logs

aldbr committed:
add -ddd system for all elements

aldbr committed:
add -ddd system for all elements

aldbr committed:
add color on logging

aldbr committed:
add color on logging

aldbr committed:
add multi line support

aldbr committed:
add multi line support

aldbr committed:
add some tests on setLevel

aldbr committed:
add some tests on setLevel

aldbr committed:
add stdout, stderr, file backend support 

via dirac.cfg
to be continued...

aldbr committed:
add stdout, stderr, file backend support 

via dirac.cfg
to be continued...

aldbr committed:
improve genericity of the logger with the different backends

aldbr committed:
improve genericity of the logger with the different backends

aldbr committed:
add a new BackendOptions section in cfg 

in order to improve the code

aldbr committed:
add a new BackendOptions section in cfg 

in order to improve the code

aldbr committed:
add support of the LogShowLine option

aldbr committed:
add support of the LogShowLine option

aldbr committed:
remove "verbose script" part code

aldbr committed:
remove "verbose script" part code

aldbr committed:
improve exception log display

aldbr committed:
improve exception log display

aldbr committed:
Merge pull request #1 from aldbr/integration

Update

aldbr committed:
New branch

Develop the solution: replace gLogger by a wrapper

aldbr committed:
New branch

Develop the solution: replace gLogger by a wrapper

aldbr committed:
New branch 

Include the gLogger wrapper

aldbr committed:
New branch 

Include the gLogger wrapper

aldbr committed:
Fix server connection problem

aldbr committed:
Fix server connection problem

aldbr committed:
Fix protected calls problem from Logger

aldbr committed:
Fix protected calls problem from Logger

aldbr committed:
Fix getLogger() problem

aldbr committed:
Fix getLogger() problem

aldbr committed:
Delete calls to protected attributes of gLogger

aldbr committed:
Delete calls to protected attributes of gLogger

aldbr committed:
Add old levels through the gLogger wrapper

aldbr committed:
Add old levels through the gLogger wrapper

aldbr committed:
Nothing special

aldbr committed:
Nothing special

aldbr committed:
Fix handler bug

aldbr committed:
Fix handler bug

aldbr committed:
Create a singleton: LoggingConfiguration

aldbr committed:
Create a singleton: LoggingConfiguration

aldbr committed:
Fix debug level at the initialization

aldbr committed:
Fix debug level at the initialization

aldbr committed:
Move the configuredLogging variable in LoggingConfiguration

aldbr committed:
Move the configuredLogging variable in LoggingConfiguration

aldbr committed:
Some little improvements

aldbr committed:
Some little improvements

aldbr committed:
Add LogLevel functionality to get log level from cfg file

And add comments to all methods

aldbr committed:
Add LogLevel functionality to get log level from cfg file

And add comments to all methods

aldbr committed:
Delete the custom logger class

aldbr committed:
Delete the custom logger class

aldbr committed:
Delete Exception custom level and  the exception custom display

aldbr committed:
Delete Exception custom level and  the exception custom display

aldbr committed:
Remove prefix on each line of a multiple line message

aldbr committed:
Remove prefix on each line of a multiple line message

aldbr committed:
Remove the custom threadID

aldbr committed:
Remove the custom threadID

aldbr committed:
Create a returned boolean in level methods

aldbr committed:
Create a returned boolean in level methods

aldbr committed:
Change level order : inverse Warn and Notice

aldbr committed:
Change level order : inverse Warn and Notice

aldbr committed:
Change back level order

aldbr committed:
Change back level order

aldbr committed:
Some changes in the structure

- LoggingConfiguration -> gLogging : a kind of logging wrapper
- Backends = wrapper of handler + format

aldbr committed:
Some changes in the structure

- LoggingConfiguration -> gLogging : a kind of logging wrapper
- Backends = wrapper of handler + format

aldbr committed:
Replace section "NewBackendsOptions" by "BackendOptions" in gLogging

aldbr committed:
Replace section "NewBackendsOptions" by "BackendOptions" in gLogging

aldbr committed:
Change LogLevels: static class

aldbr committed:
Change LogLevels: static class

aldbr committed:
Some improvements on the backend structure and protect attributes

aldbr committed:
Some improvements on the backend structure and protect attributes

aldbr committed:
Improve the structure of the BaseFormatter class

aldbr committed:
Improve the structure of the BaseFormatter class

aldbr committed:
Give a copy of the config for each backend

aldbr committed:
Give a copy of the config for each backend

aldbr committed:
Change comments to follow the comment rules

aldbr committed:
Change comments to follow the comment rules

aldbr committed:
Fix pylint errors and warnings

aldbr committed:
Fix pylint errors and warnings

aldbr committed:
Create an equivalent to the remote backend

aldbr committed:
Create an equivalent to the remote backend

aldbr committed:
Improve RemoteHandler

aldbr committed:
Improve RemoteHandler

aldbr committed:
Create some new test in Framework, delete old classes in TestLogger

aldbr committed:
Create some new test in Framework, delete old classes in TestLogger

aldbr committed:
Add some tests : Display options, SubLogger

aldbr committed:
Add some tests : Display options, SubLogger

aldbr committed:
Add some tests on the system name and the log name

aldbr committed:
Add some tests on the system name and the log name

aldbr committed:
Improve TestLogRecordCreation

aldbr committed:
Improve TestLogRecordCreation

aldbr committed:
Improve TestDisplayOptions

aldbr committed:
Improve TestDisplayOptions

aldbr committed:
Improve tests creating a mother test class

aldbr committed:
Improve tests creating a mother test class

aldbr committed:
Improve TestLevels

aldbr committed:
Improve TestLevels

aldbr committed:
Improve TestSubLogger

aldbr committed:
Improve TestSubLogger

aldbr committed:
Add RCSid on some files

aldbr committed:
Add RCSid on some files

aldbr committed:
Merge remote-tracking branch 'origin/loggerIntegration-v2' into loggerIntegration-v2

chaen commented on FrameworkSystem/Client/Logger.py:
```diff
@@ -1,8 +1,15 @@
+from DIRAC.FrameworkSystem.private.standardLogging.LoggerWrapper import LoggerWrapper
+
+#old logger
+oldgLogger = Logger()
```
You can remove that line, as well as the import

aldbr replied:
I use the _oldgLogger_ variable in my tests, is this not fine ?

chaen replied:
Maybe add a comment there.

aldbr replied:
Done.

chaen commented on FrameworkSystem/private/logging/Logger.py:
```diff
@@ -147,7 +147,7 @@ def getAllPossibleLevels( self ):
   def shown( self, levelName ):
```
is Logger.py still used somewhere ?
Why do you modify it ?

aldbr replied:
Yes, I use Logger.py in my tests, and I use _shown_ but I notice a problem with the method, so I corrected it.

chaen commented on FrameworkSystem/private/standardLogging/Backend/Backend.py:
```diff
@@ -0,0 +1,53 @@
+
+__RCSID__ = "$Id$"
+
+from abc import ABCMeta, abstractmethod
```
What is that ? :)

aldbr replied:
I wanted to create an abstract class, is it not correct ?

chaen replied:
You mean to act just as an interface ? Don't bother doing things like this. Just call it AbstractBackend, and return an error or an exception in the methods ;-)

aldbr replied:
More as an abstract class, with the possibility to add code in some methods to factorize without anyone can use them. 
If I transform it as an interface, I have to delete the code in the getHandler and setFormat methods, and I have to duplicate this code in each subclass. Are you sure ?

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/Backend/Backend.py:
```diff
@@ -0,0 +1,53 @@
+  """
+  __metaclass__ = ABCMeta
+
+  def __init__(self, handler, formatter):
```
Docstring is missing

aldbr replied:
Done.

chaen commented on FrameworkSystem/private/standardLogging/Backend/FileBackend.py:
```diff
@@ -0,0 +1,36 @@
+from DIRAC.FrameworkSystem.private.standardLogging.Formatter.BaseFormatter import BaseFormatter
+
+
+class FileBackend(Backend):
```
I suppose that ```FileBackend``` is to write the output to a file ? If so, please just state it in the doc

aldbr replied:
Yes. It is done.

chaen commented on FrameworkSystem/private/standardLogging/Backend/FileBackend.py:
```diff
@@ -0,0 +1,36 @@
+
+  def __init__(self):
+    super(FileBackend, self).__init__(None, BaseFormatter)
+    self.__fileName = 'Dirac-log_%s.log' % getpid()
```
Was it already like this in gLogger ?

aldbr replied:
Yes, is it a problem ? 
I add a comment on this point too.

chaen commented on FrameworkSystem/private/standardLogging/Backend/FileBackend.py:
```diff
@@ -0,0 +1,36 @@
+"""
```
Just for me to understand: ```FileBacked``` just configure logging to use the FileHandler and a specific BaseFormater ?

aldbr replied:
Yes, exactly. I added comments on this point.

chaen commented on FrameworkSystem/private/standardLogging/Backend/RemoteBackend.py:
```diff
@@ -0,0 +1,42 @@
+class RemoteBackend(Backend):
+  """
+  File backend wrapper
+  FileHandler() + BaseFormatter
```
Incorrect comments

aldbr replied:
I changed the comment.

chaen commented on FrameworkSystem/private/standardLogging/Backend/RemoteBackend.py:
```diff
@@ -0,0 +1,42 @@
+"""
+ServerBackend wrapper
```
I don't understand the difference between ```RemoteBackend``` and ```RemoteHandler```

aldbr replied:
I added a comment on this point. 
RemoteBackend is the link between the handler and the formatter
RemoteHandler is a custom handler to send log messages to a remote service of DIRAC
Is it clear ?

chaen commented on FrameworkSystem/private/standardLogging/Backend/RemoteBackend.py:
```diff
@@ -0,0 +1,42 @@
+
+  def __init__(self):
+    super(RemoteBackend, self).__init__(None, BaseFormatter)
+    self.__site = None
```
comments

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/Backend/StderrBackend.py:
```diff
@@ -0,0 +1,30 @@
+import sys
+
+from DIRAC.FrameworkSystem.private.standardLogging.Backend.Backend import Backend
+from DIRAC.FrameworkSystem.private.standardLogging.Formatter.BaseFormatter import BaseFormatter
```
We had agreed to use color for stderr as well

aldbr replied:
Yes, I thought it was a problem before, but with Backend objects, I can easily keep the distinction between stdout and stderr, so it is not a problem to not use color for stderr. 
Do you want to use color for stderr ?

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,50 @@
+    # pre treatment
+    logname = record.name
+
+    if record.name == "root":
```
I am sure you can factorize all these lines much better ;-)

aldbr replied:
It is done.

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,50 @@
+
+    stringRecord = super(BaseFormatter, self).format(record)
+
+    # this is important because logger creates the record and the same
```
I don't undertand the comment sorry. logger or logging ?

aldbr replied:
I changed it, I hope it is better.

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,50 @@
+    """
+    Overriding
+
+    Format for DIRAC use: add the "system/component" to the logname.
```
docstring

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/Formatter/ColoredBaseFormatter.py:
```diff
@@ -0,0 +1,60 @@
+from DIRAC.FrameworkSystem.private.standardLogging.Formatter.BaseFormatter import BaseFormatter
+
+
+class ColoredBaseFormatter(BaseFormatter):
```
Isn't there already a color formater in logging?

aldbr replied:
No, there is no color formatter in logging.

chaen commented on FrameworkSystem/private/standardLogging/Handler/RemoteHandler.py:
```diff
@@ -0,0 +1,94 @@
+"""
```
I don't understand what this class does. Is it on the client side ? Is it a server ? Is it the same as the previous system or not ?

aldbr replied:
It is almost exactly the same as the old RemoteBackend. This class send log messages to the DIRAC service "SystemLogging". 
I changed comments, do you understand now ?

chaen replied:
yep, thanks

chaen commented on FrameworkSystem/private/standardLogging/LogLevels.py:
```diff
@@ -0,0 +1,58 @@
+  @classmethod
+  def getLevelValue(cls, sName):
+    """
+    :return: a level value according to a level name
```
docstring incomplete

aldbr replied:
I added some comments and completed dostrings.

chaen commented on FrameworkSystem/private/standardLogging/LogLevels.py:
```diff
@@ -0,0 +1,58 @@
+    """
+    sName = sName.upper()
+    result = None
+    if cls.__levelDict.has_key(sName):
```
```has_key``` is deprecated. Use ```in```

Btw, you can use the ```get``` method of the dictionary to factorize your lines

```
result = None
if cls.__levelDict.has_key(sName):
  result - cls.__levelDict[sName]
return result
```
is equivalent to
``` 
return cls.__levelDict.get(sName)
```
(look at the get documentation)

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/LogLevels.py:
```diff
@@ -0,0 +1,58 @@
+    for lev in cls.__levelDict:
+      if cls.__levelDict[lev] == level:
+        return lev
+    return "Unknown"
```
I would say this deserves an exception. It should never ever happen

aldbr replied:
I don't understand why. I replace "Unknown" by None at the moment.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+
+import logging
+
+from DIRAC.FrameworkSystem.private.standardLogging.LoggingWrapper import LoggingWrapper
```
I am confused between ```LoggingWrapper``` and ```LoggerWrapper```.

aldbr replied:
I added some comments on LoggerWrapper, I hope it will be clear now.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+
+  __gLogging = LoggingWrapper()
+
+  def __init__(self, fathername='', name=''):
```
Default should rather be ```None``` (good practice)

chaen replied:
And add docstring

aldbr replied:
I can't put them at None, because the empty string '' correspond to a logger name. 
None can function for the root logger but not for children so it is complicated.
I added a docstring where I tried to explain this.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+      self.__logger = logging.getLogger(fathername).getChild(name)
+
+  def initialized(self):
+    logging.warn("Initialized: Deleted method.")
```
What is this ? Just deprecated method ?

aldbr replied:
Yes, I changed all deprecated methods in verbose with the message "Deleted method. Do not use it". Is it fine ?

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+    logging.warn("Initialized: Deleted method.")
+    return True
+
+  def showHeaders(self, yesno=True):
```
Not your fault, but just to say it: yesno is the stupidest name possible for a boolean !!!!!! :D

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+    LoggerWrapper.__gLogging.showThreadIDs(yesno)
+
+  def registerBackends(self, desiredBackends):
+    logging.warn("registerBackends: Deleted method. Logging register its backends itself.")
```
I am not sure I understand the consequences of that. Who was calling 'registerBackends' before ? And how does logging know what it has to register now ? We can discuss such points face to face Thursday

aldbr replied:
In DIRAC and lhcbDIRAC, no one use this now. But it is a public method, may be someone use this in other projects, it is just a security. Yes we will discuss about this Thursday.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+
+  def initialize(self, systemName, cfgPath):
+    """
+    Configure gLogger depending on the component used.
```
gLogger or logging ?

aldbr replied:
the root logger of logging, I changed this comment.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+    """
+    Always level
+    """
+    level = LogLevels.getLevelValue('ALWAYS')
```
as a general comment, I think we said it would be nice to have variables we can use like ```gLogger.DEBUG``` to get the level

aldbr replied:
Yes, it would be nice but there are many use of string level in the existing code: 
about 110 calls to setLevel in DIRAC and 57 in lhcbDIRAC, so it is complicated.

chaen commented on FrameworkSystem/private/standardLogging/LoggerWrapper.py:
```diff
@@ -0,0 +1,208 @@
+    return result
+
+  def showStack(self):
+    logging.warn("showStack: Deleted method.")
```
also a general remark: we do not want to see all these messages, especially not in WARN level. 
We can potentially put it in verbose

aldbr replied:
I changed to verbose.

chaen commented on FrameworkSystem/private/standardLogging/LoggingWrapper.py:
```diff
@@ -0,0 +1,181 @@
+
+  __instance = None
+
+  def __new__(cls):
```
I thought you did not need to do that because you are creating it as a class attribute of gLogger ?

aldbr replied:
Yes but it is still necessary to avoid that someone creates a new LoggingWrapper in the package. We will discuss about that Thursday if you want.

chaen commented on FrameworkSystem/private/standardLogging/LoggingWrapper.py:
```diff
@@ -0,0 +1,181 @@
+    """
+    # initialization of the default parameters
+    logging.getLogger().setLevel(LogLevels.getLevelValue('NOTICE'))
+    logging.Formatter.converter = time.gmtime
```
Important note: we want UTC time

aldbr replied:
Yes, we have UTC time, do not worry :-)

chaen commented on FrameworkSystem/private/standardLogging/LoggingWrapper.py:
```diff
@@ -0,0 +1,181 @@
+
+    # initialization of the backends
+    self.__backendsList = []
+    self.__backendsDict = {'stdout': StdoutBackend(),
```
This is not very flexible, but I think for a first version, we might consider it ok

aldbr replied:
I am not sure to understand "not very flexible" .

chaen commented on FrameworkSystem/private/standardLogging/LoggingWrapper.py:
```diff
@@ -0,0 +1,181 @@
+    """
+    for stringHandler in listHandler:
+      stringHandler = stringHandler.lower()
+      stringHandler = stringHandler.strip(" ")
```
you do not need to specify ```" "``` since it is the default

aldbr replied:
Ok, it is done.

chaen commented on FrameworkSystem/private/standardLogging/LoggingWrapper.py:
```diff
@@ -0,0 +1,181 @@
+          self.__backendsList.append(backend)
+      else:
+        self.__updateFormat()
+        logging.warning("Unexistant method for showing messages Unexistant %s logging method", stringHandler)
```
I don't understand the comment

aldbr replied:
I changed it.

chaen commented on FrameworkSystem/test/testLoggerWrapper/tests/TestDisplayOptions.py:
```diff
@@ -0,0 +1,157 @@
+import unittest
+import thread
+
+# sut
```
sut ? :)

aldbr replied:
I have no idea of what is that, I removed it :-)

chaen commented on FrameworkSystem/test/testLoggerWrapper/tests/TestLevels.py:
```diff
@@ -0,0 +1,208 @@
+    self.assertEqual(oldgLogger.shown('info'), False)
+    self.assertEqual(oldgLogger.shown('warn'), False)
+    
+    self.assertEqual(oldgLogger.shown('notice'), True)
```
Why is there differences here and bellow between the new and the old one ?

aldbr replied:
In the old gLogger, 'notice' and 'error' are at the same level -30 and 30. So when I set the level to 'error', I have also 'notice' messages. 
In the new gLogger, as the level system is linear, notice is under error. So if I set the level to 'error' I have no 'notice' message. Do you understand ?

chaen commented on FrameworkSystem/test/testLoggerWrapper/tests/TestLogRecordCreation.py:
```diff
@@ -0,0 +1,328 @@
+    - old gLogger: === Exception === ...
+    """
+    try:
+      badIdea = 1 / 0
```
:-)

chaen reviewed (commented):
So in general, very good work ! :-)
A few general comments:
* TestLoggerSystem should go. There is no reason to commit that. It however might be interesting to keep theses examples in the documentations. Not as such, but just snippets of code to show the differences
* There is a big lack of documentation. There should be more developer documentations. You have put efforts for adding docstrings and so, which is very nice. But this is not enough. The docstring explains the api, and what it does. But you should add more comment inside to explain what/why/how you are doing. Finally, there should be a documentation in the docs/developers section to explain how to use it from the client point of view (there you can put the snippets), and you should explain the overall architecture of your system (how you wrap logging)
* In general, the names of your class are confusing. I think we already talked about it. For example I don't understand the difference between gLoggerWrapper and gLoggingWrapper. (sorry if you explained me already, I forgot :-) ) That is also where the documentation is useful :-) 

But again in general, very good work ! 
We can discuss complicated things when I am back on Thursday if you like

aldbr committed:
Add a comment on the oldgLogger presence

aldbr committed:
Delete TestLoggerSystem

aldbr committed:
Update gitignore

aldbr committed:
Juste remove some useless lines

aldbr committed:
Add more comments in Backend.py

aldbr committed:
Add comments in FileBackend.py

aldbr committed:
Add more comments in RemoteBackend.py

aldbr committed:
Add more comments in StderrBackend and StdoutBackend

aldbr committed:
Improve BaseFormatter.py

aldbr committed:
Add some comments in ColorBaseFormatter.py

aldbr committed:
Add more comments in RemoteHandler.py

aldbr committed:
Improve LogLevels.py

aldbr committed:
Add some comments in LogLevels.py

aldbr committed:
Add some comments in LoggerWrapper + change deprecated method messages

aldbr committed:
Some improvements in LoggerWrapper.py

aldbr committed:
Add some comments in LoggerWrapper.py

aldbr committed:
Add some comments in LoggingWrapper.py

aldbr committed:
Remove a useless comment

aldbr committed:
Improve getSubName method in LoggerWrapper.py

aldbr committed:
Change Backend to AbstractBackend

aldbr committed:
Add color to stderr

chaen commented on .gitignore:
```diff
@@ -55,3 +55,5 @@ pytests.xml
 coverage.xml
 Local_*
+
+TestLoggerSystem/
```
That shouldn't be here ;-)

aldbr replied:
I removed it. What is the reason exactly ?

aldbr committed:
Replace warn messages by comments and add the showStack method

aldbr committed:
Modify names and deprecated methods information in comments

 LoggingWrapper -> loggingInitializer
 LoggerWrapper -> gLogging

aldbr committed:
Remove LoggingInitializer

+ add registerBackends for all gLogging
+ add showHeaders showThreadIDs for all gLogging owning a backend: inherit from the gLogging parent and can be modified after.

aldbr committed:
Give to showHeaders and showThreadID the same behaviour as setLevel

aldbr committed:
Separate gLogging in gLogging and gLoggingRoot

aldbr committed:
Remove unused imports

aldbr committed:
Delete stdout backend from the root gLogging when it is initialized

chaen commented on FrameworkSystem/private/standardLogging/Backend/AbstractBackend.py:
```diff
@@ -0,0 +1,70 @@
+
+class AbstractBackend(object):
+  """
+  Backend is used to create an abstraction of handler and formatter concepts from logging.
```
AbstractBackend

aldbr replied:
Done

aldbr committed:
Abandon native logger level, create a level attribute based on backends

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,61 @@
+    # format the record name removing "root", replacing "." by "/"
+    record.name = record.name.replace("root", "")
+
+    if record.name != "":
```
Why do you test ? If ```record.name == ""```, then the next line will have no effect anyway

aldbr replied:
Oops, you are totally right :-)

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,61 @@
+
+    # restore the record name with the copy
+    # useful because the same record name is used by different handlers and it does not be changed.
+    record.name = logname
```
Sorry I am not sure I understand what is happening here :-) What do you mean ```It does not be changed``` ?

aldbr replied:
I changed it a little bit. 
Actually, a log record is made from a logger, it contains the information of the message.
This log record is used by all the handlers of the logger, so we do not want that the log record name change between two uses. Do you understand ?

chaen commented on FrameworkSystem/private/standardLogging/LogLevels.py:
```diff
@@ -0,0 +1,71 @@
+
+  It is useful to make conversion string-integer. 
+  In fact, logging use only integers while the oldgLogger used strings, so we need a converter.
+  Example: log.setLevel(45) in logging become log.setLevel("always") in gLogger. 
```
in practice, you would do ```log.setLevel(logging.ALWAYS)```, you never use the actual number

aldbr replied:
Yes, of course. I changed it

chaen commented on FrameworkSystem/private/standardLogging/LogLevels.py:
```diff
@@ -0,0 +1,71 @@
+    :params sName: string representing a level name
+    :return: a level value according to a level name
+    """
+    sName = sName.upper()
```
why 2 times ? :)

chaen replied:
also you do not need to store it in a variable, use it directly in the get call

aldbr replied:
Oops, I changed it :-)

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+  because it has to wrap the old gLogger.  
+  """
+
+  _componentName = "Framework"
```
what is that ?

aldbr replied:
I added a comment on its presence. 
It is the default name for the system/component name that you can find in the log messages. Do you understand ?

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    Initialization of the gLogging object.
+    :params fathername: string representing the name of the father logger in the chain.
+    :params name: string representing the name of the logger in the chain. 
+    :params father: gLogging father of this new gLogging.
```
The order of the parameters is not correct compared to the signature of the method

aldbr replied:
Ok, I changed it

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    string corresponds to the root logger. 
+
+    Example: 
+    logging.getLogger('') == logging.getLogger('root') root logger
```
```== root logger``` ?

aldbr replied:
I changed it

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+
+    # dictionary of the option state, modified by the user or not
+    # this is to give to the options the same behaviour that the logging setLevel()
+    self._optionsmodified = {'showHeaders': False, 'showThreads': False}
```
We use camel style, so it should me ```_optionsModified```
Also, I don't understand what this line has to do with ```setLevel``` ?

aldbr replied:
I changed to camel style and I added a comment to explain the idea. Do you understand now ?

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    """
+    for stringHandler in desiredBackends:
+      stringHandler = stringHandler.lower()
+      stringHandler = stringHandler.strip()
```
you can chain these calls, like ```stringHandler = stringHandler.lower().strip()```

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+        # check if the backend is already in the list
+        if backend not in self._backendsList:
+          if backendOptions is not None:
+            backend.setParameters(backendOptions.copy())
```
why copy ?

aldbr replied:
I added a little comment. 
It is to avoid that the backends modify the original "backendOptions" dictionary, it is just a safety

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    self._options[optionName] = value
+    self._optionsmodified[optionName] = True
+    self._setChildrensDisplayOptions(optionName, self._options)
+    self._updateFormat()
```
What does this call do ?

aldbr replied:
updateFormat() ? I added a little comment 
It is to create a new formatter object for the handler in the backend object, in order to modify the display of the log messages when an option changes. Do you understand ?

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    """
+    return LogLevels.getLevel(self._logger.getEffectiveLevel())
+
+  def shown(self, levelName):
```
is this used ? It will not make much sense anymore if you can change the threashold on a per backend basis

aldbr replied:
I changed it yesterday, I still have some modifications to do on this point.

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    :return: the name of the logger
+    """
+    name = self._logger.name
+    name = name.replace("root", "")
```
Can't you change the name of the logger once and for all ?

aldbr replied:
I can't because the name qualifies the logger, so if I change the name, I change the logger. I tried to change it before, it was inconclusive. 
But I created a name attribute that stores the subName of the logger. It is more simple now

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    for child in self._childrens:
+      child._setChildrensDisplayOptions(optionName, options)
+
+  def getAllPossibleLevels(self):
```
is this really used ?

aldbr replied:
getAllPossibleLevels() ? Just in one case in Interfaces/API/Job in the setLogLevel() method

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    """
+    :return: "system name/component name"
+    """
+    return gLogging._componentName
```
How can that work ? It should always return the same then ?

aldbr replied:
What is the problem ? Yes it always returns the same value

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    """
+    result = self._childExists(subName)
+    if result is None:
+      child = gLogging(self, self._logger.name, subName)
```
be careful, the name is conflicting with the paramater

aldbr replied:
I don't understand the problem

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    :params name: the name of the child
+    :return: boolean, True if it exists, else False
+    """
+    for child in self._childrens:
```
Instead of using an array for ```self._children``` (no ```s``` btw, children is already plural ;-) ), you were using a dictionary, all this would be easier no ?

aldbr replied:
Yes, of course. I changed it

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    return None
+
+  def initialized(self):
+    # initialized: Deleted method. Do not use it.
```
docstring rather than simple comment

aldbr replied:
Done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    return True
+
+  def processMessage(self, messageObject):
+    # processMessage: Deleted method. Do not use it.
```
docstring rather than simple comment

aldbr replied:
Done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,335 @@
+    return False
+
+  def flushAllMessages(self, exitCode=0):
+    # flushAllMessages: Deleted method. Do not use it.
```
docstring rather than simple comment

aldbr replied:
Done

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+
+class gLoggingRoot(gLogging):
+  """
+  gLoggingRoot is a gLogging object and it is particular because it is the first parent of the chaine.
```
if I understand well, we should never instantiate a gLogging object ?

aldbr replied:
Yes, gLoggingRoot is a singleton and represents gLogger. 
And gLogging objects are created by gLoggingRoot.

chaen replied:
Then you should say in gLogging that it should not be instantiated directly.
In fact, maybe the name is misleading, we will discuss it

aldbr replied:
I changed them to Logging and LoggingRoot

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+    """
+    super(gLoggingRoot, self).__init__()
+    self._logger = logging.getLogger('')
+    self._logger.setLevel(LogLevels.getLevelValue('DEBUG'))
```
by default you set it to ```DEBUG``` ? That's a bit aggressive no ? :-) Or was it the case in the old one ?

aldbr replied:
Actually, it is the level of the logger object, in order to send all the log messages to the central log server in the future. Now the level of the gLogging object is managed by an attribute of gLogging named _level.

aldbr replied:
I added a comment to explain this point

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+    :params systemName: string represented as "system name/component name"
+    :params cfgPath: string of the cfg file path
+    """
+    from DIRAC.ConfigurationSystem.Client.Config import gConfig
```
why do you put the import here ?

aldbr replied:
Because I can't put in at the top of the module. It creates a dependency loop I think. It was the same problem in the old gLogger.

chaen replied:
then comment ;-)

aldbr replied:
Ok, it is done

aldbr committed:
Fix little problem in the RemoteBackend

aldbr committed:
Little improvement

aldbr committed:
Add a little comment

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+  we created a singleton to keep it unique. 
+  """
+
+  __configuredLogging = False
```
What is this variable ?

aldbr replied:
It prevents the multiple configuration of the gLoggingRoot object. Actually, I don't know if it is very useful, it was from the old gLogger.

chaen replied:
Well, check ;-)

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+      gLogging._componentName = systemName
+
+      # Remove stdout from the list of backends
+      del self._backendsList[:]
```
you are removing everything here no ?

aldbr replied:
Yes, it was the case in the old gLogger. It allows the user to not print logs in any backend if he wants. I don't know if it is useful. What do you think ?

chaen replied:
Well, why not, but then the comment is wrong, you should correct it

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+      del self._backendsList[:]
+      self._logger.removeHandler(self._logger.handlers[0])
+      # Backend options
+      desiredBackendsStr = gConfig.getValue("%s/LogBackends" % cfgPath, 'stdout')
```
try 
```gConfig.getValue("%s/LogBackends" % cfgPath, ['stdout'])```
I think if you do that, it automatically splits with ```,```
But test it

aldbr replied:
It works, thanks :-)

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+      self._options['Path'] = gConfig.getValue("%s/LogShowLine" % cfgPath, False)
+
+      currentLevelName = logging.getLevelName(logging.getLogger().getEffectiveLevel())
+      levelname = gConfig.getValue("%s/LogLevel" % cfgPath, currentLevelName)
```
why don't you just test if the value is defined or not ? It is easier than getting the current level and resetting it no ?

aldbr replied:
Yes, I changed it

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+
+      retDict = gConfig.getOptionsDict("%s/BackendsOptions" % cfgPath)
+      if retDict['OK']:
+        backends = (desiredBackends, retDict['Value'])
```
Do you have several backend ? Why is it plural ?

aldbr replied:
Yes, one gLogging object can have several backends. It was the case with the old gLogger too. Did not you know ?

aldbr committed:
Change .gitignore

aldbr committed:
Change a comment in AbstractBackend

aldbr committed:
Remove an useless condition test

aldbr committed:
Add one comment line to pinpoint a code line

aldbr committed:
Change one comment in LogLevels

aldbr committed:
Little method factorization in LogLevels

aldbr committed:
Add a comment on the componentName variable presence

aldbr committed:
Change the order of one parameter in a docstring

aldbr committed:
Fix cfg file problem when logLevel is not precised

+ Change some variables to camelCase
+ Fix a comment

aldbr committed:
Precise a comment

aldbr committed:
Add one little comment

aldbr committed:
Factorize some lines

aldbr committed:
Add a little comment

aldbr committed:
Improve getSubName

aldbr committed:
Add one little comment

aldbr committed:
changes children list to a dictionary

aldbr committed:
Change comments to docstrings

aldbr committed:
Add a comment to the logger level

aldbr committed:
Factorize some lines in gLoggingRoot

aldbr committed:
Simplify the code in gLoggingRoot

aldbr committed:
Prepare the arrival of the central system logging

aldbr committed:
Prepare the arrival of the central logging server

aldbr committed:
Fix level backends after registryBackends calls

chaen commented on FrameworkSystem/private/standardLogging/Backend/RemoteBackend.py:
```diff
@@ -0,0 +1,57 @@
+    :params parameters: dictionary of parameters. ex: {'FileName': file.log}
+    """
+    self.__site = DIRAC.siteName()
+    if 'Interactive' in parameters:
```
Just do ```self.__interactive = parameters.get('Interactive')```
Same bellow :-)

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/Backend/RemoteBackend.py:
```diff
@@ -0,0 +1,57 @@
+    self._handler.setLevel(LogLevels.getLevelValue('ERROR'))
+
+  def setLevel(self, level):
+    pass
```
You should add a comment here explaining that nothing will happen because it is hardcoded

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/Backend/StderrBackend.py:
```diff
@@ -0,0 +1,36 @@
+  def __init__(self):
+    super(StderrBackend, self).__init__(None, ColoredBaseFormatter)
+
+  def setParameters(self, parameters):
```
Comment on why we pass. Nothing to set ?

aldbr replied:
Ok. it is done and yes nothing to set here

chaen commented on FrameworkSystem/private/standardLogging/Formatter/BaseFormatter.py:
```diff
@@ -0,0 +1,60 @@
+    :params record: the log record containing all the information about the log message: name, level, threadid...
+    """
+    # create a record name copy
+    logname = record.name
```
I still don't understand that. We will discuss it

aldbr replied:
Ok, the format method does not exist anymore

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+
+  # componentName is a class variable because the component name is the same for every gLogging objects
+  # its default value is "Framework" but it can be configured and composed by the system name and the component name
+  _componentName = "Framework"
```
I am not sure to understand how this works either, we will discuss it

aldbr replied:
Ok, I added some comments on this point

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+  # its default value is "Framework" but it can be configured and composed by the system name and the component name
+  _componentName = "Framework"
+
+  def __init__(self, father=None, fathername='', name=''):
```
Camel case, so it should be ```fatherName```

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+    self._parent = father
+
+    # all the different backends
+    self._backendsDict = {'stdout': StdoutBackend(),
```
This should be a class variable, it does not depend of the instance

chaen replied:
Also you can put the name in capital to show that it is a constant

aldbr replied:
Ok it is done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+  def registerBackends(self, desiredBackends, backendOptions=None):
+    """
+    Attach a list of backends to the gLogging object.
+    :params desiredBackends: a list of different names attaching to differents backends.
```
I suppose the names have to be the same than in _backendsDict ? if so, please say it explicitely

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+    Attach a list of backends to the gLogging object.
+    :params desiredBackends: a list of different names attaching to differents backends.
+                             example : ['stdout', 'file', 'server']
+    :params backendOptions: a dictionary of different backend options. 
```
So if I undertand correctly, you have one list with all the backends, and a single dic which contains options for all the backends ? That is a bit dangerous no ?

aldbr replied:
I don't know how to do this in a better way

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+    :params backendOptions: a dictionary of different backend options. 
+                            example: {'FileName': '/tmp/log.txt'}
+    """
+    for stringHandler in desiredBackends:
```
why stringHandler ? I don't understand the name

aldbr replied:
I changed it

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+        if backend not in self._backendsList:
+          if backendOptions is not None:
+            # give a copy to avoid that the backends modify the dictionary
+            backend.setParameters(backendOptions.copy())
```
It is a bit counter intuitive to have to set the parameters, and then call configure so that the parameter is applied. Was it like this in the past already ?

aldbr replied:
Ok, I merged the two methods in one called "createHandler(parameters=None)"

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+
+  def setLevel(self, levelName):
+    """
+    Set a level to the logger.
```
Update the docstring to say that it proopagaets to the Backends

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+    # - propagation from the parent to the children when their levels are not set by the developer himself
+    # - stop the propagation when a developer set a level to a child
+    self._optionsModified = {'showHeaders': False, 'showThreads': False}
+    self._levelModified = False
```
This will not work no ? 
If for example I set the level of gLogging, then create a subLogger, and then reset the level of gLogging, the level will not be propagated to the subLogger

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+
+    :return: boolean representing the result of the log record creation
+    """
+    self._logger.log(level, "%s %s", sMsg, sVarMsg, exc_info=exc_info, extra={'componentname': self.getName()})
```
Can you write comment on that ? What are these parameters ? 
Also, camel case so ```componentName```.
Also, if you put sMsg and sVarMsg in the same string, how can you send them in separate string to the Remote server ?

aldbr replied:
I writed a comment on that
I think it is not a good idea because it corresponds to a Log Record attributes and they are not in camel case. What do you think ? 
I created a new extra attribute for the sVarMsg, so they are totally separated :-)

chaen replied:
wrote, not writed ;-)
Sounds good to me

chaen commented on FrameworkSystem/private/standardLogging/gLogging.py:
```diff
@@ -0,0 +1,370 @@
+    """
+    return self.debug('')
+
+  def _updateFormat(self):
```
I though that doing this was the role of the Formater ?

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+    """
+    super(gLoggingRoot, self).__init__()
+    self._logger = logging.getLogger('')
+    # this level is not the gLogging level, it is only used to send all log messages to the central logging system
```
Here you can maybe give better explanation, and say that we put the logging to debug so that the backends can choose whatever level they want

aldbr replied:
Ok, it is done

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+    self._logger.setLevel(LogLevels.getLevelValue('DEBUG'))
+
+    # initialization of the UTC time
+    logging.Formatter.converter = time.gmtime
```
Here as well, why do you have a formater, I thought it is in the backend only ?

aldbr replied:
According to the documentation of 'time': 
"Convert a time expressed in seconds since the epoch to a struct_time in UTC in which the dst flag is always zero. "
https://docs.python.org/2/library/time.html
So it is UTC, isn't it ?

chaen commented on FrameworkSystem/private/standardLogging/gLoggingRoot.py:
```diff
@@ -0,0 +1,132 @@
+
+      # Remove stdout from the list of backends
+      del self._backendsList[:]
+      self._logger.removeHandler(self._logger.handlers[0])
```
This is also making big assumptions

aldbr replied:
I changed it, it is more clear now

aldbr committed:
Factorize code in File and Remote Backends

aldbr committed:
Change class names gLogging -> Logging

aldbr committed:
Add a comment on the import in the middle of the code

aldbr committed:
Precise a comment on the remove of stdout in LoggingRoot

aldbr committed:
Add docstring to comment "pass" methods

aldbr committed:
Add some precision on the componentName comment

aldbr committed:
change fathername to fatherName

aldbr committed:
Create backend only if the Logging needs it

Remove the limit to one backend type per Logging: now we can create many same backends for a same Logging

aldbr committed:
Change stringHandler to backendName

aldbr committed:
Add a comment on the possible values for desiredBackends

aldbr committed:
Merge setParameters and configureHandler in createHandler method

aldbr committed:
Fix the level and the option propagation

aldbr committed:
Add a new attribute to log records: varmessage

aldbr committed:
Add some comments on the log line

aldbr committed:
Add varmessage in RemoteHandler

aldbr committed:
Add a docstring comment in Logging

aldbr committed:
Add a comment on the DEBUG level of the root logger

aldbr committed:
Improve the comment

aldbr committed:
Removed BaseFormatter format method. Improve the cost of the display

Add a comment on UTC time

aldbr committed:
Add a comment on gmtime

aldbr committed:
Add a comment on non camel case

aldbr committed:
Move the format creation in AbstractBackend

aldbr committed:
Fix some comments

aldbr committed:
Add a comment

aldbr commented on FrameworkSystem/private/standardLogging/Logging.py:
```diff
@@ -0,0 +1,388 @@
+    # 'root' representing the root logger name is removed
+    # all "." separator are replaced by "/"
+    self.entireName = self._logger.name.replace("root", "")
+    self.entireName = self.entireName.replace(".", "/")
```
@chaen
Here you can see the modification of the logger name stored in self.entireName.
This name is used in the _createLogRecord method where I put it in the extra dictionary with "customname" as key.
This extra attribute is used in the display as the log name instead of the original name.

chaen replied:
Okay great. I just went through the code of logging, and there is no way to change the dot indeed. 
However, I recommend that you send an email to the mailing list, asking what they think about having a dot instead of a slash. Maybe it simplifies our life.
Btw, note that with this technique, you have a problem if the logger is called "a.b.c" no ?

aldbr committed:
Improve Logging: now accept and keep names composed by "."

aldbr committed:
Change variable names 'showHeaders' and 'showThreads'

aldbr committed:
Change updateFormat name by generateBackendsFormat

Remove some 'useless' calls to copy()

aldbr committed:
Change values() bu itervalues() in the loops

aldbr committed:
Improve _setLevel method: less expensive

aldbr committed:
Add some comments in a docstring

aldbr committed:
Improve _setDisplayOptions: less expensive

aldbr committed:
Add documentation

aldbr committed:
Improve gLogger.rst

aldbr committed:
Add svg files

aldbr committed:
Replace svg files by png files

aldbr committed:
Add images

aldbr committed:
Add a new image

aldbr committed:
Add level systems image

aldbr committed:
Modify the old gLogger documentation

aldbr committed:
Separate gLogger documentation in two distinct files

aldbr committed:
Move images

aldbr committed:
Add some internal links

aldbr committed:
Fix some failed links

aldbr committed:
Add external label links

aldbr committed:
Add a link to the advanced part

aldbr committed:
Add external link to redirect to the new doc

aldbr committed:
Improve it

aldbr committed:
Improve changes

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/Changes/index.rst:
```diff
@@ -0,0 +1,281 @@
+
+::
+
+    gLogger.setLevel('notice')
```
here you can use the example we just did together. It is well descriptive

aldbr replied:
Ok, it is done

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/Changes/index.rst:
```diff
@@ -0,0 +1,281 @@
+    # stderr
+    # > VERBOSE: message
+
+Moreover, change a format on a *Logging* which has no *Backend* have no
```
Even on the children ?

aldbr replied:
The sentence was false, I removed it

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
++------------+----------------------------------------------------------------------------------------------------+
+| Error      | Used when an error occur but do not need to force the program exit.                                |
++------------+----------------------------------------------------------------------------------------------------+
+| Exception  | Actually a specification of the Error level which must be used before raising an exception.        |
```
No, it is used when an exception is trapped

aldbr replied:
Ok, I changed it

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
++------------+----------------------------------------------------------------------------------------------------+
+| Notice     | Used to provide an important information.                                                          |
++------------+----------------------------------------------------------------------------------------------------+
+| Warn       | Used when an error can occur.                                                                      |
```
Not really an error, but more a ``` potentially undesired behavior ``` ;-)

aldbr replied:
I changed it too

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
+Add the caller path name and its line
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The *LogShowLine* option is only available from the *cfg* file and
```
I did not know that. It shows the caller, or just 'Logging.py' ?

aldbr replied:
It shows always 'Logging.py'. It was the case in the old gLogger too. I don't know if it is a good idea to keep this behaviour

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
+
+As we may notice, *gLogger* has already a *stdout Backend* by default.
+
+Add a *Backend* to your *Logging*
```
Add a note that the next step is to have it plugable, so that anyone can define a new type of backend.

aldbr replied:
Ok, it is done

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
+~~~~~~~~~~~~~~~~~~
+
+*DIRAC* is composed by many micro services running in multiple
+processes. *gLogger* object is naturally different for two distinct
```
That is not true. There is rarely multiprocess. It is mostly multi thread.

aldbr replied:
Ok, I removed the first sentence.

chaen commented on docs/source/DeveloperGuide/AddingNewComponents/Utilities/gLogger/gLogger/Basics/index.rst:
```diff
@@ -0,0 +1,607 @@
+~~~~~~~~~~~~~~~~
+
+*gLogger* is based on the Python *logging* library which is completely
+thread-safe. Thus, it is also thread-safe.
```
Watch out ! That is a very daring statement ! In fact, I am not sure your code is thread safe at all !

aldbr replied:
I modified the last sentence

chaen replied:
Il faudra quand meme verifier que tu es thread-safe. Je n'en suis pas convaincu.

chaen reviewed (commented):
bravo ! Great doc

aldbr committed:
Improve _setLevel and _setOption

aldbr committed:
Change the setLevel example

aldbr committed:
Remove one ambiguous line

aldbr committed:
Change some level descriptions

aldbr committed:
Add a note on plugable backends

aldbr committed:
Multi processes and threads corrections

aldbr committed:
Add a comment in Logging, add a performance test class

aldbr committed:
Add the init file

aldbr committed:
Little modification

aldbr committed:
Add gLoggerDevelopment doc

aldbr committed:
Add a title name

aldbr committed:
Add the new doc in the index

aldbr committed:
Add a link to the gLogger documentation

aldbr committed:
Add a link at the top of gLoggerDevelopment

aldbr committed:
Delete TestSpeedLogCreation

aldbr committed:
Delete init

aldbr committed:
Optimize code

aldbr committed:
Correct some comments

aldbr committed:
Remove _childExists method

aldbr committed:
Remove the showLogLine option

aldbr committed:
Remove showLogLine from the documentation

aldbr committed:
Remove global oldgLogger

aldbr committed:
Remove global oldgLogger

aldbr committed:
Comment the new logging system
